### ISSUE TITLE
Texturing untextured surfaces with nearby Colors within the iOS App

### ISSUE THREAD

**Issue by electrum-bowie** [introlab/rtabmap#1358]
![image](https://github.com/user-attachments/assets/46a26533-c96f-4e83-9fe0-93cbba64fad1)
![image](https://github.com/user-attachments/assets/ca356430-4b93-4bb7-8cbc-67cfe8ef3dca)

The vertex mesh assembling mode does this but unfortunately the textured parts are extremely low quality, and as you may know, not having textures on the model itself makes it incompatible with basic software.

**Comment by electrum-bowie** [comment]
Also, if you add something like this, please keep the mesh enclosing feature ( hole filling functionality ) available, RTABMap is the only iOS software that provides actual complete 360 models, which even closes bigger voids in the capture that the LiDAR Scanner didn't reach !

It would be amazing to have a setting that can maximize the hole filling algorithm strength in the future as well, knowing that it can still leave some big gaps open. 

That combined with the requested feature in this post, additionally texturing meshes based on nearby colors, would finally enable full environmental scanning directly from the iPhone !

It would finally allow revisiting complete captures in VR, without having colorless or big voids in the models !

**Comment by matlabbe** [comment]
Thanks for the suggestion. I'll check we can do that for next ios version release.

**Comment by electrum-bowie** [comment]
Yup, using the PC to export a scan from a standalone device ( iOS ) to another standalone device ( Meta Quest VR ) isn't feasible.

**Comment by electrum-bowie** [comment]
Does the new version include this feature ?

**Comment by matlabbe** [comment]
Not yet.

**Comment by DenisDaniel707** [comment]
This feature would be a game changer for sure! The app is already a powerhouse. Hats off to @matlabbe 🙏

**Comment by matlabbe** [comment]
To add to the discussion, it seems that we can add vertex color in OBJ format, though it is not standard, so not sure about the support in software you may use. Ref: https://paulbourke.net/dataformats/obj/colour.html

If it works, that would be the easiest/fastest way to do it, as we should already know the closest color of each vertex in the mesh before texturing it.

**Comment by matlabbe** [comment]
The new version with this new feature has been released to App Store.

You can enable this export mode by changing the following settings under Settings->Assembling:

![Image](https://github.com/user-attachments/assets/4be625ff-1ddc-4cbc-b03b-602162ad519b)

This makes the mesh fully closed:
![Image](https://github.com/user-attachments/assets/9efd5071-5eb1-411e-aa74-9436fd665911)

Note that there are two modes on how we attribute vertex colors:

![Image](https://github.com/user-attachments/assets/998ab566-a740-40da-9d8d-7efb7c21e591)

By clicking on the Texture/Color blend visualization option, we better see the difference:

![Image](https://github.com/user-attachments/assets/4070caaf-c508-4c6f-8b1c-be3259b06fe3)

Depending on the software you are using, this is how the meshes may look like. To get the previous image with no seams, we have a custom shader to not apply vertex color when there is a texture.

When we export to OBJ, if we enabled vertex color, the color will be added to each vertex.
